plipy/ksvd_cdl.py

```diff
@@ -0,0 +1,333 @@
+import time
+
+import numpy as np
+import pandas as pd
+import scipy
+import numba
+
+
+@numba.jit((numba.float64[:, :], numba.float64[:, :]), cache=True)
+def compute_DtD(u, v):  # pragma: no cover
+    n_atoms, n_times_atom = v.shape
+
+    DtD = np.zeros(shape=(n_atoms, n_atoms, 2 * n_times_atom - 1))
+    t0 = n_times_atom - 1
+    for k0 in range(n_atoms):
+        for k in range(n_atoms):
+            for t in range(n_times_atom):
+                if t == 0:
+                    DtD[k0, k, t0] = np.dot(v[k0], v[k])
+                else:
+                    DtD[k0, k, t0 + t] = np.dot(v[k0, :-t], v[k, t:])
+                    DtD[k0, k, t0 - t] = np.dot(v[k0, t:], v[k, :-t])
+    DtD *= np.dot(u, u.T).reshape(n_atoms, n_atoms, 1)
+    return DtD
+
+
+class KSVD_CDL():
+    """
+    K-SVD for convolutional dictionary laerning with one rank constraint.
+
+    Parameters
+    ----------
+    n_atoms : int
+        Number of atoms
+    n_times_atom : int
+        Size of the convolutional kernels
+    n_iter : int, optional
+        Number of iterations, default to 10
+    tol : float, optional
+        Threshold for sparse coding error, default to 0.1
+    max_iter : int, optional
+        Maximum number of sparse coding iterations, should not be reached if
+        tol is reached before.
+    correlation : str {'full', 'fast', 'debug'}
+        Method to update the correlation in the sparse coding. If full, update
+        the residual and recompute the full correlation each time, as done in
+        the original algorithm. If fast, update directly the correlation using
+        DtD. If 'debug', check that the two methods match.
+    """
+    def __init__(self, n_atoms, n_times_atom, n_iter=10, tol=0.1,
+                 max_iter=100000, correlation='fast', random_seed=42):
```

Comment:
Avoid hard coded parameter that would surprise a user.

```suggestion
                 max_iter=100000, correlation='fast', random_seed=None):
```